File: core/clientLaya/game/src/shine/net/httpRequest/BaseHttpRequest.ts
namespace Shine
{
	export class BaseHttpRequest extends BaseData
	{
		/** www对象 */
		private _www:Laya.HttpRequest;
	
		/** 地址 */
		protected _url:string;
	
		/** httpMethod */
		protected _method:number=HttpMethodType.Get;
	
		/** post数据 */
		protected _postStream:BytesWriteStream;
	
		/** 结果数据 */
		protected _resultStream:BytesReadStream;
	
	    /* 超时时间,默认30秒 */
	    public timeOut: number = 30 * 1000;
	
	    /** 是否处理 */
	    private _dided: boolean = false;
	
	    constructor()
	    {
			super();
	    }
	
	    /** 把数据拷贝下 */
	    protected copyData(): void  
	    {
	    }
	
	    /* 发送 */
	    public send(): void 
	    {
	        this.write();
	
	        this._www=new Laya.HttpRequest();

	        this._www.once(Laya.Event.COMPLETE, this, this.preComplete);
			this._www.once(Laya.Event.ERROR, this, this.preError);
	
	        if (this._method == HttpMethodType.Get) 
	        {
	            this._www.send(this._url, null, 'get', 'text');
	        }
	        else if (this._method == HttpMethodType.Post)
	        {
				var arr:ArrayBuffer=this._postStream.getByteArray();

				if(ShineSetting.clientUseBase64)
				{
					this._www.send(this._url,Base64.encode(arr) , 'post', 'arraybuffer');//['content-type', 'pplication/octet-stream']
				}
				else
				{
					this._www.send(this._url,arr , 'post', 'arraybuffer');
				}
	        }
	
	        NetControl.addHttpRequest(this);
	    }
	
		protected write():void
	    {
	
	    }
	
	    /** 是否已结束 */
	    public isDone():boolean
	    {
	        return this._dided;
	    }
	
		/** 析构 */
		public dispose():void
	    {
	        if (this._dided)
	            return;
	
	        this._dided = true;
	
	        this.onError();
	
	    }
	
		/** 预完成 */
		private preComplete():void
	    {
	        if (this._dided)
	            return;
	
	        this._dided = true;
	
	        this._resultStream = new BytesReadStream(this._www.data);
	
	        this.read();
	
	        this.onComplete();
	    }
	
	    private preError():void
	    {
	        if (this._dided)
	            return;
	
	        this._dided = true;
	
	        this.onError();
	    }
	
		protected read():void
	    {
	
	    }
	
		/** IO出错 */
		protected onError():void
	    {
	
	    }
	
		/** 完成*/
		protected onComplete():void
	    {
	
	    }
	}
}

File: core/client/game/src/shine/net/httpRequest/BaseHttpRequest.cs
using UnityEngine;

namespace ShineEngine
{
	/// <summary>
	/// 基础http请求
	/// </summary>
	[Hotfix(needChildren = false)]
	public class BaseHttpRequest:BaseData
	{
		/** www对象 */
		private WWW _www;

		/** 地址 */
		protected string _url;

		/** httpMethod */
		protected int _method=HttpMethodType.Get;

		/** post数据 */
		protected BytesWriteStream _postStream;

		/** 结果数据 */
		protected BytesReadStream _resultStream;

		/// <summary>
		/// 超时时间,默认30秒
		/// </summary>
		public int timeOut=30 * 1000;

		/** 是否处理 */
		private bool _dided=false;

		public BaseHttpRequest()
		{
		}

		/** 把数据拷贝下 */
		protected virtual void copyData()
		{
		}

		/// <summary>
		/// 发送
		/// </summary>
		public void send()
		{
			write();

			if(_method==HttpMethodType.Get)
			{
				//WWWForm f = new WWWForm();
				_www=new WWW(_url);
			}
			else if(_method==HttpMethodType.Post)
			{
				_www=new WWW(_url,_postStream.getByteArray());
			}

			NetControl.addHttpRequest(this);
		}

		protected virtual void write()
		{
		}

		/// <summary>
		/// 是否完成
		/// </summary>
		public bool isDone()
		{
			return _www.isDone;
		}

		/** 超时 */
		public void onTimeOut()
		{
			dispose();
		}

		/// <summary>
		/// 析构
		/// </summary>
		public void dispose()
		{
			if(_dided)
				return;

			_dided=true;

			onError();

			if(_www!=null)
			{
				_www.Dispose();
				_www=null;
			}
		}

		/** 预完成 */
		public void preComplete()
		{
			if(_dided)
				return;

			_dided=true;

			if(!string.IsNullOrEmpty(_www.error))
			{
				onError();
			}
			else
			{
				_resultStream=new BytesReadStream(_www.bytes);

				read();

				onComplete();
			}

			if(_www!=null)
			{
				_www.Dispose();
				_www=null;
			}
		}

		protected virtual void read()
		{

		}

		/// <summary>
		/// IO出错
		/// </summary>
		protected virtual void onError()
		{

		}

		/// <summary>
		/// 完成
		/// </summary>
		protected virtual void onComplete()
		{

		}
	}
}